<!-- HTML header for doxygen 1.8.4-->
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.4"/>
<title>NVBIO: NVBIO</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="extra_stylesheet.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript">
    (function(i, s, o, g, r, a, m) {
        i['GoogleAnalyticsObject'] = r; i[r] = i[r] || function() {
            (i[r].q = i[r].q || []).push(arguments)
        }, i[r].l = 1 * new Date(); a = s.createElement(o),
  m = s.getElementsByTagName(o)[0]; a.async = 1; a.src = g; m.parentNode.insertBefore(a, m)
    })(window, document, 'script', '//www.google-analytics.com/analytics.js', 'ga');
    ga('create', 'UA-47310325-1', 'nvlabs.github.io');
    ga('send', 'pageview');
</script>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">NVBIO
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.4 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li class="current"><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="modules.html"><span>Modules</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="examples.html"><span>Examples</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
</div><!-- top -->
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(6)"><span class="SelectionMark">&#160;</span>Typedefs</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(7)"><span class="SelectionMark">&#160;</span>Enumerations</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(8)"><span class="SelectionMark">&#160;</span>Enumerator</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(9)"><span class="SelectionMark">&#160;</span>Friends</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(10)"><span class="SelectionMark">&#160;</span>Macros</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(11)"><span class="SelectionMark">&#160;</span>Groups</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(12)"><span class="SelectionMark">&#160;</span>Pages</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="headertitle">
<div class="title">NVBIO </div>  </div>
</div><!--header-->
<div class="contents">
<div class="textblock"> 
 <img src="nvidia_cubes.png" style="position:relative; bottom:-10px; border:0px;"/>
 &nbsp;&nbsp;
<dl class="section user"><dt></dt><dd><br/>
 <a href="https://github.com/NVlabs/nvbio">NVBIO</a> is a library of reusable components designed by <a href="http://www.nvidia.com">NVIDIA</a> to accelerate bioinformatics applications using <em>CUDA</em>. Though it is specifically designed to unleash the power of <em>NVIDIA</em> <b>GPU</b>s, most of its components are completely cross-platform and can be used both from host C++ and device CUDA code. </dd></dl>
<dl class="section user"><dt></dt><dd>The purpose of NVBIO is twofold: it can be thought of both as a solid basis to build new, modern applications targeting GPUs, which deferring the core computations to a library will always automatically and transparently benefit from new advances in GPU computing, and as example material to design novel bioinformatics algorithms for massively parallel architectures. </dd></dl>
<dl class="section user"><dt></dt><dd>Additionally, NVBIO contains a suite of applications built on top of it, including a re-engineered implementation of the famous <a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">Bowtie2</a> short read aligner. Unlike many prototypes, <a class="el" href="nvbowtie_page.html">nvBowtie</a> is an attempt to build an industrial strength aligner, reproducing most of Bowtie2's original features as well as adding a few more, such as efficient support for direct BAM (and soon CRAM) output. </dd></dl>
<dl class="section user"><dt></dt><dd>Similarly, it contains the fastest BWT builders for strings and string-sets available to date. NVBIO can in fact perform the BWT of individual strings containing several billion characters or string-sets containing up to ~100 billion symbols at an unprecedented speed of roughly <em>80M symbols/s</em>, on a single Tesla K40.</dd></dl>
<h1><a class="anchor" id="Features"></a>
Features</h1>
<dl class="section user"><dt></dt><dd>Here's a short list of some of the most noteworthy features of NVBIO: </dd></dl>
<dl class="section user"><dt></dt><dd><ul>
<li>many <a class="el" href="strings_page.html">strings</a> and <a class="el" href="strings_page.html#StringSetInterface">string-sets</a> representations, including <a class="el" href="hello_dna_page.html">packed string</a> encodings for arbitrary <a class="el" href="strings_page.html#AlphabetsSection">alphabets</a></li>
<li><a class="el" href="fmindex_page.html">FM-index</a> and <a class="el" href="fmindex_page.html#BidirectionalFMIndex">Bidirectional FM-index</a> construction and lookup algorithms on arbitrary strings, and compile-time alphabets.</li>
<li>novel, state-of-the-art parallel <a class="el" href="sufsort_page.html">suffix sorting</a> and <a class="el" href="sufsort_page.html">BWT</a> construction algorithms for very large strings and string sets, on arbitrary alphabets</li>
<li><a class="el" href="waveletfm_page.html">Wavelet Trees</a> and other succint data structures for large alphabet strings</li>
<li>efficient parallel <a class="el" href="fmindex_page.html#MEMFiltersSection">k-Maximal Extension Match</a> filters</li>
<li>novel, state-of-the-art parallel <a class="el" href="qgram_page.html">Q-Gram</a> (or k-mer) <a class="el" href="qgram_page.html">Index</a> construction and lookup</li>
<li><a class="el" href="alignment_page.html">Dynamic Programming Alignment</a> with parallel implementations of most scoring algorithms (e.g. Edit-Distance, Smith-Waterman, Gotoh,...) with support for global, semi-global or local alignment, in both banded and full-matrix modes, on arbitrary string types</li>
<li>parallel <a class="el" href="bloom_filter_page.html">Bloom Filter</a> construction and lookup</li>
<li>state-of-the-art support for many <a class="el" href="primitives_page.html">parallel primitives</a>, like sorting, reductions, prefix-sums, stream compaction, run-length encoding, etc</li>
<li>state-of-the-art <a class="el" href="io_page.html">I/O</a> of common sequencing formats (e.g. FASTA, FASTQ, BAM), with transparent, parallel compression/decompression </li>
</ul>
</dd></dl>
<dl class="section user"><dt></dt><dd>importantly, all of the above algorithms and data-structures have both CPU and GPU implementations, easily selected through template specialization.</dd></dl>
<h1><a class="anchor" id="Links"></a>
Links</h1>
 
 <div>
 <a href="https://github.com/NVlabs/nvbio"><img src="browse-icon.png" style="position:relative; bottom:-10px; border:0px;"/></a>
 &nbsp;&nbsp;
 <a href="https://github.com/NVlabs/nvbio">Browse or fork NVBIO at GitHub</a>
 <br>

 <a href="https://groups.google.com/forum/#!forum/nvbio-users"><img src="groups-icon.png" style="position:relative; bottom:-10px; border:0px;"/></a>
 &nbsp;&nbsp;
 <a href="https://groups.google.com/forum/#!forum/nvbio-users">The NVBIO users forum</a>
 <br>

 <a href="https://github.com/NVlabs/nvbio"><img src="download-icon.png" style="position:relative; bottom:-10px; border:0px;"/></a>
 &nbsp;&nbsp;
 <a href="https://github.com/NVlabs/nvbio">Download NVBIO 1.0</a>
 </div>
<h1><a class="anchor" id="Documentation"></a>
Documentation</h1>
<dl class="section user"><dt></dt><dd>Documentation for the NVBIO suite can be found here: <br/>
<ul>
<li><a class="el" href="nvbio_page.html">NVBIO</a> - the NVBIO library</li>
<li><a class="el" href="nvbowtie_page.html">nvBowtie</a> - a re-engineered implementation of the famous <a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">Bowtie2</a> short read aligner</li>
<li><a class="el" href="nvbwt_page.html">nvBWT</a> - a tool to perform BWT-based reference indexing</li>
<li><a class="el" href="nvssa_page.html">nvSSA</a> - a tool to build auxiliary Sampled Suffix Arrays needed for reference indexing</li>
<li><a class="el" href="nvfm_server_page.html">nvFM-server</a> - a shared memory FM-index server</li>
<li><a class="el" href="nvsetbwt_page.html">nvSetBWT</a> - a tool to perform BWT-based read indexing</li>
</ul>
</dd></dl>
<h1><a class="anchor" id="RecentNew"></a>
Recent News</h1>
<dl class="section user"><dt></dt><dd><table class="doxtable">
<tr>
<td style="white-space: nowrap; vertical-align:text-top;">24/1/2015 <br/>
 NVBIO 1.0 </td><td style="vertical-align:text-top;" width="90%"><ul>
<li><b>New Features:</b> <br/>
<ul>
<li>Major stability and <b>specificity</b> and <b>sensitivity</b> improvements to <a class="el" href="nvbowtie_page.html">nvBowtie</a> <div class="image">
<img src="nvBowtie-ROC-pe-100-zoom.png"  style="position:relative; bottom:-10px; border:0px;" width="50%" height="50%"/>
</div>
</li>
<li>Added support for finding and reporting discordant alignments</li>
<li>Improved direct <b>BAM output speed</b> by over an order of magnitude with transparent parallel compression  </li>
</ul>
</li>
</ul>
</td></tr>
</table>
</dd></dl>
<dl class="section user"><dt></dt><dd><table class="doxtable">
<tr>
<td style="white-space: nowrap; vertical-align:text-top;">13/1/2015 <br/>
 NVBIO 0.9.99 </td><td style="vertical-align:text-top;" width="90%"><ul>
<li><b>New Features:</b> <br/>
<ul>
<li>Made nvBowtie's command line compatible with Bowtie2</li>
<li>Added many missing features to nvBowtie, including support for:<ul>
<li>trimming (-3/&ndash;trim3, -5/&ndash;trim5)</li>
<li>forward/reverse alignment (&ndash;nofw,&ndash;norc)</li>
<li>presets (&ndash;fast,&ndash;very-fast,&ndash;sensitive,&ndash;very-sensitive)  </li>
</ul>
</li>
</ul>
</li>
</ul>
</td></tr>
</table>
</dd></dl>
<dl class="section user"><dt></dt><dd><table class="doxtable">
<tr>
<td style="white-space: nowrap; vertical-align:text-top;">20/12/2014 <br/>
 NVBIO 0.9.98 </td><td style="vertical-align:text-top;" width="90%"><ul>
<li><b>New Features:</b> <br/>
<ul>
<li>Added multi-GPU support to nvBowtie, together with many accuracy and sensitivity improvements</li>
<li>Released a new version of nvSetBWT based on our freshly developed massively parallel algorithm: <a href="http://arxiv.org/pdf/1410.0562.pdf">http://arxiv.org/pdf/1410.0562.pdf</a>  </li>
</ul>
</li>
</ul>
</td></tr>
</table>
</dd></dl>
<dl class="section user"><dt></dt><dd><table class="doxtable">
<tr>
<td style="white-space: nowrap; vertical-align:text-top;">15/09/2014 <br/>
 NVBIO 0.9.97 </td><td style="vertical-align:text-top;" width="90%"><ul>
<li><b>New Features:</b> <br/>
<ul>
<li>Added many Bowtie2-compatible command line options to nvBowtie</li>
<li>Added &ndash;ungapped-mates option to nvBowtie, which doubles speed at a minor cost in sensitivity (&lt;0.2%)</li>
</ul>
</li>
<li><b>Bug Fixes:</b> <br/>
<ul>
<li>Fixed several memory-corruption bugs in nvBowtie's paired-end alignment  </li>
</ul>
</li>
</ul>
</td></tr>
</table>
</dd></dl>
<dl class="section user"><dt></dt><dd><table class="doxtable">
<tr>
<td style="white-space: nowrap; vertical-align:text-top;">26/05/2014 <br/>
 NVBIO 0.9.90 </td><td style="vertical-align:text-top;" width="90%"><ul>
<li><b>New Features:</b> <br/>
<ul>
<li>Added a new <a class="el" href="strings_page.html#AlphabetsSection">Alphabets</a> module for managing different alphabets.</li>
<li>Added a new <a class="el" href="sequence_io_page.html">Sequence IO</a> module for managing all kinds of sequence data; unlike the old io::ReadData, the containers in this module support various types of alphabet encodings, and can be used both for loading reads as well as for reading or mapping large reference data.</li>
<li>Rewrote the <a class="el" href="fmindex_io_page.html">FM-Index IO</a> module; the new io::FMIndexData offers more uniform interfaces, and unlike its predecessor no longer holds any reference data, which can now be separately loaded through the new <a class="el" href="sequence_io_page.html">Sequence IO</a> module.</li>
<li>Rewrote the <a class="el" href="group___batch_scheduler.html">Batch Alignment Schedulers</a>, adding support for a new OpenMP host backend.  </li>
</ul>
</li>
</ul>
</td></tr>
</table>
</dd></dl>
<dl class="section user"><dt></dt><dd><table class="doxtable">
<tr>
<td style="white-space: nowrap; vertical-align:text-top;">16/05/2014 <br/>
 NVBIO 0.9.7 </td><td style="vertical-align:text-top;" width="90%"><ul>
<li><b>New Features:</b> <br/>
<ul>
<li>Simplified the interfaces for <a class="el" href="packed_streams_page.html">Packed Streams</a></li>
<li>Added a suite of easy-to-use and highly efficient host / device parallel primitives</li>
<li>More uniform handling of <a class="el" href="structnvbio_1_1vector.html">nvbio::vector</a> views</li>
</ul>
</li>
<li><b>Bug Fixes:</b> <br/>
<ul>
<li>Fixed option parsing bug in nvBWT.  </li>
</ul>
</li>
</ul>
</td></tr>
</table>
</dd></dl>
<dl class="section user"><dt></dt><dd><table class="doxtable">
<tr>
<td style="white-space: nowrap; vertical-align:text-top;">08/05/2014 <br/>
 NVBIO 0.9.6 </td><td style="vertical-align:text-top;" width="90%"><ul>
<li><b>New Features:</b> <br/>
<ul>
<li>Added a set of step-by-step introductory tutorials</li>
<li>Sped up MEM filtering  </li>
</ul>
</li>
</ul>
</td></tr>
</table>
</dd></dl>
<dl class="section user"><dt></dt><dd><table class="doxtable">
<tr>
<td style="white-space: nowrap; vertical-align:text-top;">15/04/2014 <br/>
 NVBIO 0.9.5 </td><td style="vertical-align:text-top;" width="90%"><ul>
<li><b>New Features:</b> <br/>
<ul>
<li>Added a whole new <a class="el" href="qgram_page.html">Q-Gram Module</a> for q-gram indexing and q-gram counting.</li>
<li>Added a banded <a class="el" href="alignment_page.html#AlignersAnchor">Myers bit-vector</a> edit-distance algorithm running just short of <b>1 TCUPS</b>!</li>
<li>Added parallel primitives for simple and efficient <a class="el" href="strings_page.html#SeedingAnchor">seed extraction</a> from string and string-sets</li>
<li>Added parallel <a class="el" href="fmindex_page.html#FMIndexFilters">FM-index</a> and <a class="el" href="fmindex_page.html#MEMFilters">MEM</a> filters</li>
<li>Further improved GPU BWT construction throughput by ~20%.</li>
<li>Added a set of examples showing how to build <a class="el" href="qgram_page.html">q-gram</a> and <a class="el" href="fmindex_page.html">FM-index</a> based all-mappers in a few lines of code, as well as how to extract all <a class="el" href="fmindex_page.html#MEMFilters">MEM</a>s from a set of reads</li>
</ul>
</li>
<li><b>Bug Fixes:</b> <br/>
<ul>
<li>Fixed access violation bug in ModernGPU's SegmentedSort (thanks Sean Baxter), used at the core of our suffix sorting routines</li>
<li>Fixed FMIndexDataDevice constructor bug when loading the reverse index  </li>
</ul>
</li>
</ul>
</td></tr>
</table>
</dd></dl>
<h1><a class="anchor" id="Dependencies"></a>
Dependencies</h1>
<dl class="section user"><dt></dt><dd>NVBIO depends on the following external libraries:<ul>
<li><a href="http://nvlabs.github.io/cub/">CUB</a></li>
<li><a href="https://sites.google.com/site/yuta256/">SAIS</a></li>
<li><a href="http://www.zlib.net/">zlib</a></li>
<li><a href="http://www.barrgroup.com/Embedded-Systems/How-To/CRC-Calculation-C-Code">crc</a></li>
<li>a modification of Nathaniel McClatchey's <a href="https://github.com/nmcclatchey/Priority-Deque/">priority_deque</a></li>
</ul>
</dd></dl>
<h1><a class="anchor" id="Requirements"></a>
Requirements</h1>
<dl class="section user"><dt></dt><dd>NVBIO has been designed for GPUs supporting at least CUDA's <em>Compute Capability 3.5</em>. Due to the high memory requirements typical of bionformatics algorithms, Tesla K20, K20x or K40 are recommended.</dd></dl>
<h1><a class="anchor" id="Licensing"></a>
Licensing</h1>
<dl class="section user"><dt></dt><dd>NVBIO has been developed by <a href="http://www.nvidia.com">NVIDIA Corporation</a> and is licensed under BSD.</dd></dl>
<h1><a class="anchor" id="Contributors"></a>
Contributors</h1>
<dl class="section user"><dt></dt><dd>The main contributors of NVBIO are <a href="#" onclick="location.href='mai'+'lto:'+'jpa'+'nt'+'ale'+'on'+'i@n'+'vi'+'dia'+'.c'+'om'; return false;">Jacopo Pantaleoni</a> and <a href="#" onclick="location.href='mai'+'lto:'+'nsu'+'bt'+'il@'+'nv'+'idi'+'a.'+'com'; return false;">Nuno Subtil</a>.</dd></dl>
 
 <a href="http://research.nvidia.com"><img src="cuda_small.png" style="position:relative; bottom:-10px; border:0px;"/></a>
 &nbsp;&nbsp;
 </div></div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on Fri Jan 30 2015 10:26:03 for NVBIO by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.4
</small></address>
</body>
</html>
