<!-- HTML header for doxygen 1.8.4-->
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.4"/>
<title>NVBIO: nvBowtie</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="extra_stylesheet.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript">
    (function(i, s, o, g, r, a, m) {
        i['GoogleAnalyticsObject'] = r; i[r] = i[r] || function() {
            (i[r].q = i[r].q || []).push(arguments)
        }, i[r].l = 1 * new Date(); a = s.createElement(o),
  m = s.getElementsByTagName(o)[0]; a.async = 1; a.src = g; m.parentNode.insertBefore(a, m)
    })(window, document, 'script', '//www.google-analytics.com/analytics.js', 'ga');
    ga('create', 'UA-47310325-1', 'nvlabs.github.io');
    ga('send', 'pageview');
</script>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">NVBIO
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.4 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="modules.html"><span>Modules</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="examples.html"><span>Examples</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(6)"><span class="SelectionMark">&#160;</span>Typedefs</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(7)"><span class="SelectionMark">&#160;</span>Enumerations</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(8)"><span class="SelectionMark">&#160;</span>Enumerator</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(9)"><span class="SelectionMark">&#160;</span>Friends</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(10)"><span class="SelectionMark">&#160;</span>Macros</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(11)"><span class="SelectionMark">&#160;</span>Groups</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(12)"><span class="SelectionMark">&#160;</span>Pages</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="index.html">NVBIO</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="headertitle">
<div class="title">nvBowtie </div>  </div>
</div><!--header-->
<div class="contents">
<div class="textblock"> 
   <img src="nvidia_cubes.png" style="position:relative; bottom:-10px; border:0px;"/>
  <dl class="section user"><dt></dt><dd><br/>
 nvBowtie is a GPU-accelerated re-engineering of Bowtie2, a very widely used short-read aligner. While being completely rewritten from scratch, nvBowtie reproduces many (though not all) of the features of Bowtie2.</dd></dl>
<h1><a class="anchor" id="PerformanceSection"></a>
Performance</h1>
<dl class="section user"><dt></dt><dd>nvBowtie is designed to exploit all the massive parallelism of modern GPUs, thus enabling a much higher alignment throughput at equal accuracy, or higher accuracy in the same time. Here's a graph showing nvBowtie's performance compared to bowtie2 on an Illumina HiSeq 2000 dataset (the first 10M reads of ERR161544) and an IonProton run, using both end-to-end and local alignment. The alignment results show 99.98% agreement at high MAPQ. All bowtie2 tests were run using 20 CPU threads, and default aligment options:</dd></dl>
<div class="image">
<img src="benchmark-nvbowtie-speedup.png"  style="position:relative; bottom:-10px; border:0px;" width="80%" height="80%"/>
</div>
<h1><a class="anchor" id="SpecificitySection"></a>
Specificity and Sensitivity</h1>
<dl class="section user"><dt></dt><dd>While targeting maximum performance, nvBowtie is designed to match Bowtie2 as closely as possible, with the explicit goal of mantaining the same specificity and sensitivity characteristics. The following ROC curves have been generated using the excellent online testing platform <a href="http://www.bioplanet.com/gcat">GCAT</a>. As can be seen, nvBowtie matches Bowtie2's results to within extreme accuracy for a wide range of read types and error distributions. The only noticeable (and still very tiny) differences are concentrated around very low mapping quality alignments.</dd></dl>
<div class="image">
<img src="nvBowtie-ROC-se-100.png"  style="position:relative; bottom:-10px; border:0px;" width="95%" height="95%"/>
</div>
 <div class="image">
<img src="nvBowtie-ROC-pe-100.png"  style="position:relative; bottom:-10px; border:0px;" width="95%" height="95%"/>
</div>
 <div class="image">
<img src="nvBowtie-ROC-pe-250.png"  style="position:relative; bottom:-10px; border:0px;" width="95%" height="95%"/>
</div>
<h1><a class="anchor" id="nvBowtieArch"></a>
Architecture</h1>
<dl class="section user"><dt></dt><dd>In order to take advantage of the massive parallelism available in modern processor architectures, nvBowtie re-implements the same underlying algorithms as Bowtie2 taking a fundamentally different approach. In fact, while Bowtie2 is essentially designed to operate on a single read at a time (possibly having multiple CPU threads working on a different read), carrying the entire alignment process in what is basically a very complex chain of nested function calls, nvBowtie works at all times with large batches of reads, and treats their alignment as a complex pipeline composed by many relatively simple but deeply parallel stages. In many of these stages parallelism is spread at a much finer granularity than at the reads-level, for example by processing many candidate hits from each read at the same time.</dd></dl>
<dl class="section user"><dt></dt><dd>There are effectively a few different pipelines, one for each kind of alignment:<ul>
<li>best mapping of single-end reads</li>
<li>best mapping of paired-end reads</li>
<li>all mapping of single-end reads</li>
<li>all mapping of paired-end reads (not implemented yet)</li>
</ul>
</dd></dl>
<h2><a class="anchor" id="BestMapping-SingleEnd"></a>
best-mapping, single-end reads</h2>
<dl class="section user"><dt></dt><dd>This pipeline can be described by the following pseudo-code:</dd></dl>
<div class="fragment"><div class="line">best-alignments = {}</div>
<div class="line"></div>
<div class="line">seed-mapping-queue[in] := <a class="code" href="group___c_u_d_a_primitives.html#ga9dcce72ad9a2d44f92aa508b91b3378d">all</a> reads</div>
<div class="line"><span class="keywordflow">while</span> (seed-mapping-queue[in] not empty)</div>
<div class="line">{</div>
<div class="line">    (seed-mapping-queue[out], seed-deques) := <a class="code" href="group___mapping.html#gad30b05e1249763bbda8b4745c92a42ae">map</a>( seed-mapping-queue[in] );</div>
<div class="line"></div>
<div class="line">    score( seed-mapping-queue[in], seed-deques, best-alignments );          <span class="comment">// find the best alignments for each read</span></div>
<div class="line"></div>
<div class="line">    seed-mapping-queue[in] = seed-mapping-queue[out];</div>
<div class="line">}</div>
<div class="line"><a class="code" href="group___traceback_detail.html#ga9fc78dc4580fa6a4c275ab530655a6db">traceback_best</a>( best-alignments );                                          <span class="comment">// traceback the best alignments</span></div>
</div><!-- fragment --><dl class="section user"><dt></dt><dd>where score() is a sub-pipeline:</dd></dl>
<div class="fragment"><div class="line">score([in] input-reads, [in] seed-deques, [in/out] best-alignments)</div>
<div class="line">{</div>
<div class="line">   <a class="code" href="group___work_queue_detail.html#ga7c6f0457212ffc4b696a01d5858485b6">active</a>-reads := input-reads</div>
<div class="line">   <span class="keywordflow">while</span> (<a class="code" href="group___work_queue_detail.html#ga7c6f0457212ffc4b696a01d5858485b6">active</a>-reads not empty)</div>
<div class="line">   {</div>
<div class="line">        (hits-queue, <a class="code" href="group___work_queue_detail.html#ga7c6f0457212ffc4b696a01d5858485b6">active</a>-reads) := <a class="code" href="group___basic_utils.html#gabc969419bf0a10c17e84ceedd624cd49">select</a>( <a class="code" href="group___work_queue_detail.html#ga7c6f0457212ffc4b696a01d5858485b6">active</a>-reads, seed-deques );  <span class="comment">// select the next round of hits to process</span></div>
<div class="line">        <a class="code" href="group___f_m_index.html#ga036b0d1d5e131f9ae61bf9fc37e4949f">locate</a>( hits-queue );                                               <span class="comment">// SA coordinates -&gt; linear genome coordinates conversion</span></div>
<div class="line">        <a class="code" href="group___scoring.html#gade1f1010066627cb4cf979ec7e67435f">score_best</a>( hits-queue );                                           <span class="comment">// assign a score to the selected hits using DP alignment</span></div>
<div class="line">        <a class="code" href="group___reduce.html#gaf4b932cd61fa1d6575f21d3e68523bea">score_reduce</a>( hits-queue, best-alignments );                        <span class="comment">// keep track of the best alignments found so far for each read</span></div>
<div class="line">   }</div>
<div class="line">}</div>
</div><!-- fragment --><dl class="section user"><dt></dt><dd>The various functions involved correspond to different pipeline stages:<ul>
<li><a class="el" href="group___mapping.html">Mapping</a></li>
<li><a class="el" href="group___locate.html">Locate</a></li>
<li><a class="el" href="group___select.html">Select</a></li>
<li><a class="el" href="group___scoring.html">Scoring</a></li>
<li><a class="el" href="group___reduce.html">Reduce</a></li>
<li><a class="el" href="group___traceback.html">Traceback</a></li>
</ul>
</dd></dl>
<dl class="section user"><dt></dt><dd>The data flowing through the pipeline goes mostly through a single-data structure of type ScoringQueues (<a class="el" href="group___scoring_queues_module.html">Scoring Queues</a>), which itself contains the active-reads (ScoringQueues::active_reads), the hits-queue (ScoringQueues::hits) and an index allowing to walk all the hits belonging to each read (ScoringQueues::hits_index).</dd></dl>
<h2><a class="anchor" id="BestMapping-PairedEnd"></a>
best-mapping, paired-end reads</h2>
<dl class="section user"><dt></dt><dd>This pipeline can be described by the following pseudo-code:</dd></dl>
<div class="fragment"><div class="line">best-anchor-alignments   = {}</div>
<div class="line">best-opposite-alignments = {}</div>
<div class="line"></div>
<div class="line"><span class="keywordflow">for</span> (anchor in {mate1,mate2})</div>
<div class="line">{</div>
<div class="line">    seed-mapping-queue[in] := <a class="code" href="group___c_u_d_a_primitives.html#ga9dcce72ad9a2d44f92aa508b91b3378d">all</a> reads</div>
<div class="line">    <span class="keywordflow">while</span> (seed-mapping-queue[in] not empty)</div>
<div class="line">    {</div>
<div class="line">        (seed-mapping-queue[out], seed-deques) := <a class="code" href="group___mapping.html#gad30b05e1249763bbda8b4745c92a42ae">map</a>( seed-mapping-queue[in] );</div>
<div class="line"></div>
<div class="line">        score( seed-mapping-queue[in], seed-deques, best-anchor-alignments, best-opposite-alignments );</div>
<div class="line"></div>
<div class="line">        seed-mapping-queue[in] = seed-mapping-queue[out];</div>
<div class="line">    }</div>
<div class="line">}</div>
<div class="line"><a class="code" href="group___traceback_detail.html#ga9fc78dc4580fa6a4c275ab530655a6db">traceback_best</a>( best-anchor-alignments );</div>
<div class="line"><a class="code" href="group___traceback.html#ga9e62d21577ddfffb2557d8108ebf0851">opposite_traceback_best</a>( best-opposite-alignments );</div>
</div><!-- fragment --><dl class="section user"><dt></dt><dd>where again score() is a sub-pipeline:</dd></dl>
<div class="fragment"><div class="line">score([in] input-reads, [in] seed-deques, [in/out] best-anchor-alignments, [in/out] best-opposite-alignments)</div>
<div class="line">{</div>
<div class="line">   <a class="code" href="group___work_queue_detail.html#ga7c6f0457212ffc4b696a01d5858485b6">active</a>-reads := input-reads</div>
<div class="line">   <span class="keywordflow">while</span> (<a class="code" href="group___work_queue_detail.html#ga7c6f0457212ffc4b696a01d5858485b6">active</a>-reads not empty)</div>
<div class="line">   {</div>
<div class="line">        (hits-queue, <a class="code" href="group___work_queue_detail.html#ga7c6f0457212ffc4b696a01d5858485b6">active</a>-reads) := <a class="code" href="group___basic_utils.html#gabc969419bf0a10c17e84ceedd624cd49">select</a>( <a class="code" href="group___work_queue_detail.html#ga7c6f0457212ffc4b696a01d5858485b6">active</a>-reads, seed-deques );  <span class="comment">// select the next round of hits to process</span></div>
<div class="line">        <a class="code" href="group___f_m_index.html#ga036b0d1d5e131f9ae61bf9fc37e4949f">locate</a>( hits-queue );                                               <span class="comment">// SA coordinates -&gt; linear genome coordinates conversion</span></div>
<div class="line">        <a class="code" href="group___scoring.html#gaf1229ce3a465e49c0dc7836abd37747b">anchor_score_best</a>( hits-queue );                                    <span class="comment">// assign a score to the selected hits using DP alignment</span></div>
<div class="line">        <a class="code" href="group___scoring.html#ga065c6b73e4afdce4afcd4cfac4f7badc">opposite_score_best</a>( hits-queue );                                  <span class="comment">// assign a score to opposite mate of the selected hits using DP alignment</span></div>
<div class="line">        <a class="code" href="group___reduce.html#ga156907d3ba196849440395a44ad429b0">score_reduce_paired</a>( hits-queue, best-anchor-alignments, best-opposite-alignments ); <span class="comment">// keep track of the best alignments found so far for each read</span></div>
<div class="line">   }</div>
<div class="line">}</div>
</div><!-- fragment --><h1><a class="anchor" id="nvBowtieUsage"></a>
Usage</h1>
<dl class="section user"><dt></dt><dd>At the moment, the command line options of nvBowtie differ from those of bowtie2. A comprehensive list can be obtained running:</dd></dl>
<pre class="fragment">*  ./nvBowtie --help
* 
*  nvBowtie [options]
*   options:
*     General:
*       -U                 file-name     unpaired reads
*       -1                 file-name     first mate reads
*       -2                 file-name     second mate reads
*       -S                 file-name     output file (.sam|.bam)
*       -x                 file-name     reference index
*       --verbosity                      verbosity level
*       --upto  | -u       int [-1]      maximum number of reads to process
*       --trim3 | -3       int [0]       trim the first N bases of 3'
*       --trim5 | -5       int [0]       trim the first N bases of 5'
*       --nofw                           do not align the forward strand
*       --norc                           do not align the reverse-complemented strand
*       --device           int [0]       select the given cuda device(s) (e.g. --device 0 --device 1 ...)
*       --file-ref                       load reference from file
*       --server-ref                     load reference from server
*       --phred33                        qualities are ASCII characters equal to Phred quality + 33
*       --phred64                        qualities are ASCII characters equal to Phred quality + 64
*       --solexa-quals                   qualities are in the Solexa format
*     Paired-end:
*       --ff                             paired mates are forward-forward
*       --fr                             paired mates are forward-reverse
*       --rf                             paired mates are reverse-forward
*       --rr                             paired mates are reverse-reverse
*       --minins     | -I   int [0]      minimum insert length
*       --minins     | -X   int [500]    maximum insert length
*       --overlap                        allow overlapping mates
*       --no-overlap                     disable overlapping mates
*       --dovetail                       allow dovetailing mates
*       --no-discordant                  disable discordant alignments
*       --no-mixed                       only report paired alignments
*       --ungapped-mates | -ug           perform ungapped mate alignment
*     Seeding:
*       --seed-len   | -L   int  [22]                       seed lengths
*       --seed-freq  | -i   func [S,1,local ? 0.75 : 1.15]  seed interval function
*       --max-hits          int  [100]                      maximum amount of seed hits
*       --max-reseed | -R   int  [2]                        number of reseeding rounds
*     Extension:
*       --all        | -a                perform all-mapping (i.e. find and report all alignments)
*       --local                          perform local alignment
*       --rand                           randomized seed selection
*       --no-rand                        disable randomized seed selection
*       --max-dist          int [15]     maximum edit distance
*       --max-effort-init   int [15]     initial maximum number of consecutive extension failures
*       --max-effort | -D   int [15]     maximum number of consecutive extension failures
*       --min-ext           int [30]     minimum number of extensions per read
*       --max-ext           int [400]    maximum number of extensions per read
*       --fast                           apply the fast presets
*       --very-fast                      apply the very-fast presets
*       --sensitive                      apply the sensitive presets
*       --very-sensitive                 apply the very-sensitive presets
*       --fast-local                     apply the fast presets
*       --very-fast-local                apply the very-fast presets
*       --sensitive-local                apply the sensitive presets
*       --very-sensitive-local           apply the very-sensitive presets
*     Scoring:
*       --scoring-scheme    filename     use a given scoring scheme
*       --ma                int          match bonus
*       --mp                int,int      max,min mismatch penalties
*       --np                int          N penalty
*       --rdg               int,int      read gap open / extension penalties
*       --rfg               int,int      reference gap open / extension penalties
*       --score-min         func         minimum score function
*     Reporting:
*       --mapQ-filter | -Q  int [0]      minimum mapQ threshold
*     Debug
*       --report            filename     generate an HTML report
* </pre><dl class="section user"><dt></dt><dd>While most parameters should be easy to understand, a major difference is that nvBowtie relies on an external program to build the reference indices: </dd></dl>
<dl class="section user"><dt></dt><dd><ul>
<li><a class="el" href="nvbwt_page.html">nvBWT</a> : builds the BWT indices of the reference FASTA files</li>
</ul>
</dd></dl>
<dl class="section user"><dt></dt><dd>e.g. suppose you have the human genome in a single FASTA file, <em>hg19.fa</em>; the following commands will create all indices needed by nvBowtie:</dd></dl>
<pre class="fragment">*  ./nvBWT hg19.fa hg19-index
* </pre><dl class="section user"><dt></dt><dd>At this point, one can run nvBowtie:</dd></dl>
<pre class="fragment">*  ./nvBowtie -x hg19 -U my_reads.fastq -S my_reads.bam
* </pre><dl class="section user"><dt></dt><dd>Notice that nvBowtie supports direct output of BAM files, which has been carefully optimized and parallelized in order to cope with the superior alignment throughput.</dd></dl>
<dl class="section user"><dt></dt><dd>Another noteworthy option is to let nvBowtie fetch the reference index from a <em>shared memory</em> server which can be run in the background: the <a class="el" href="nvfm_server_page.html">nvFM-server</a>. It be launched with:</dd></dl>
<pre class="fragment">*  ./nvFM-server hg19-index hg19 &amp;
* </pre><dl class="section user"><dt></dt><dd>nvBowtie can then pickup the reference from the server:</dd></dl>
<pre class="fragment">*  ./nvBowtie -x hg19 -U my_reads.fastq -S my_reads.bam
* </pre><h2><a class="anchor" id="ScoringSchemeSection"></a>
scoring schemes</h2>
<dl class="section user"><dt></dt><dd>The <em>&ndash;scoring-scheme filename</em> option allows to provide a custom Smith-Waterman scoring scheme through a text file, where each line must contain a token value pair. The tokens and default values are reported below:</dd></dl>
<pre class="fragment">*  match               0        // local alignment: 2
*  mm-penalty-min      2
*  mm-penalty-max      6
*  N-penalty-min       1
*  N-penalty-max       1
*  score-min-const     -0.6     // local alignment: 0
*  score-min-coeff     -0.6     // local alignment: 10
*  score-min-type      linear   // local alignment: log
*  N-ceil-const        0
*  N-ceil-coeff        0.15
*  read-gap-const      5
*  read-gap-coeff      3
*  ref-gap-const       5
*  ref-gap-coeff       3
*  gap-free            5
* </pre> </div></div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on Wed Jan 28 2015 12:56:16 for NVBIO by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.4
</small></address>
</body>
</html>
