<!-- HTML header for doxygen 1.8.4-->
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.4"/>
<title>NVBIO: nvbio::io::OutputFile Struct Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="extra_stylesheet.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript">
    (function(i, s, o, g, r, a, m) {
        i['GoogleAnalyticsObject'] = r; i[r] = i[r] || function() {
            (i[r].q = i[r].q || []).push(arguments)
        }, i[r].l = 1 * new Date(); a = s.createElement(o),
  m = s.getElementsByTagName(o)[0]; a.async = 1; a.src = g; m.parentNode.insertBefore(a, m)
    })(window, document, 'script', '//www.google-analytics.com/analytics.js', 'ga');
    ga('create', 'UA-47310325-1', 'nvlabs.github.io');
    ga('send', 'pageview');
</script>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">NVBIO
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.4 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="modules.html"><span>Modules</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(6)"><span class="SelectionMark">&#160;</span>Typedefs</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(7)"><span class="SelectionMark">&#160;</span>Enumerations</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(8)"><span class="SelectionMark">&#160;</span>Enumerator</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(9)"><span class="SelectionMark">&#160;</span>Friends</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(10)"><span class="SelectionMark">&#160;</span>Macros</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(11)"><span class="SelectionMark">&#160;</span>Groups</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(12)"><span class="SelectionMark">&#160;</span>Pages</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="namespacenvbio.html">nvbio</a></li><li class="navelem"><a class="el" href="namespacenvbio_1_1io.html">io</a></li><li class="navelem"><a class="el" href="structnvbio_1_1io_1_1_output_file.html">OutputFile</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Methods</a> &#124;
<a href="#pub-static-methods">Static Public Methods</a> &#124;
<a href="#pro-methods">Proteced Methods</a> &#124;
<a href="#pro-attribs">Protected Members</a> &#124;
<a href="structnvbio_1_1io_1_1_output_file-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">nvbio::io::OutputFile Struct Reference<div class="ingroups"><a class="el" href="group___output.html">Output</a></div></div>  </div>
</div><!--header-->
<div class="contents">
<a name="details" id="details"></a><h2 class="groupheader">Detailed description</h2>
<div class="textblock"><p>The output file interface.</p>
<p>This class is the base class for all file output classes. By itself, it implements a "null" output, which discards all data sent into it. Other classes derive from this to implement various file output formats.</p>
<p>After each alignment pass in the pipeline, the aligner should call <a class="el" href="structnvbio_1_1io_1_1_output_file.html#a97bc4b3558dba005cff6b3a3cf73da5e">OutputFile::process</a>. This call takes a handle to the alignment buffers in GPU memory (a <a class="el" href="structnvbio_1_1io_1_1_g_p_u_output_batch.html">GPUOutputBatch</a> structure). The implementation ensures that any data required is copied out before the call returns, allowing the caller to invalidate the GPU memory immediately.</p>
<p>Sequence alignment is generally done in batches of reads; for each batch multiple <a class="el" href="structnvbio_1_1io_1_1_output_file.html#a97bc4b3558dba005cff6b3a3cf73da5e">OutputFile::process</a> calls can be made. Two extra enum parameters serve to identify which alignment pass we're handling at each call. Each batch should be preceded by a <a class="el" href="structnvbio_1_1io_1_1_output_file.html#a637d0c9547a1c3e3aaa0c5aa2d2d3827">OutputFile::start_batch</a> call; at the end of each batch, <a class="el" href="structnvbio_1_1io_1_1_output_file.html#a0c7f1c3ab1ae5f1b975baab07806f395" title="Mark a batch of alignment results as complete. ">OutputFile::end_batch</a> should be called. In most cases, data is only written to disk after <a class="el" href="structnvbio_1_1io_1_1_output_file.html#a0c7f1c3ab1ae5f1b975baab07806f395" title="Mark a batch of alignment results as complete. ">OutputFile::end_batch</a> is called.</p>
<p>The factory method <a class="el" href="structnvbio_1_1io_1_1_output_file.html#af28e4dd0e708422eb0e29b0b11f02f2d">OutputFile::open</a> is used to create <a class="el" href="structnvbio_1_1io_1_1_output_file.html">OutputFile</a> objects. It parses the file name extension to determine the file format for the output. </p>

<p>Definition at line <a class="el" href="output__file_8h_source.html#l00061">61</a> of file <a class="el" href="output__file_8h_source.html">output_file.h</a>.</p>
</div>
<p><code>#include &lt;<a class="el" href="output__file_8h_source.html">output_file.h</a>&gt;</code></p>
<div class="dynheader">
Inheritance diagram for nvbio::io::OutputFile:</div>
<div class="dyncontent">
 <div class="center">
  <img src="structnvbio_1_1io_1_1_output_file.png" usemap="#nvbio::io::OutputFile_map" alt=""/>
  <map id="nvbio::io::OutputFile_map" name="nvbio::io::OutputFile_map">
<area href="structnvbio_1_1io_1_1_bam_output.html" alt="nvbio::io::BamOutput" shape="rect" coords="0,56,139,80"/>
<area href="structnvbio_1_1io_1_1_debug_output.html" alt="nvbio::io::DebugOutput" shape="rect" coords="149,56,288,80"/>
<area href="structnvbio_1_1io_1_1_sam_output.html" alt="nvbio::io::SamOutput" shape="rect" coords="298,56,437,80"/>
</map>
 </div></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Methods</h2></td></tr>
<tr class="memitem:adec2613fac91d4c2dc3a2a401b7b2e7d"><td class="memItemLeft" align="right" valign="top">virtual&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structnvbio_1_1io_1_1_output_file.html#adec2613fac91d4c2dc3a2a401b7b2e7d">~OutputFile</a> ()</td></tr>
<tr class="separator:adec2613fac91d4c2dc3a2a401b7b2e7d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa109737d25afa8d93a55572803aca499"><td class="memItemLeft" align="right" valign="top">virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structnvbio_1_1io_1_1_output_file.html#aa109737d25afa8d93a55572803aca499">configure_mapq_evaluator</a> (<a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> <a class="el" href="structnvbio_1_1io_1_1_map_q_evaluator.html">io::MapQEvaluator</a> *mapq, int <a class="el" href="structnvbio_1_1io_1_1_output_file.html#aac600141218c722d6f69dfe624f85fb6">mapq_filter</a>)</td></tr>
<tr class="memdesc:aa109737d25afa8d93a55572803aca499"><td class="mdescLeft">&#160;</td><td class="mdescRight">Configure the MapQ evaluator. Must be called prior to any batch processing.  <a href="#aa109737d25afa8d93a55572803aca499">More...</a><br/></td></tr>
<tr class="separator:aa109737d25afa8d93a55572803aca499"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a637d0c9547a1c3e3aaa0c5aa2d2d3827"><td class="memItemLeft" align="right" valign="top">virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structnvbio_1_1io_1_1_output_file.html#a637d0c9547a1c3e3aaa0c5aa2d2d3827">start_batch</a> (<a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> <a class="el" href="group___sequence_i_o.html#gaeda29063a8a35631e18372d8a6001e27">io::SequenceDataHost</a> *<a class="el" href="structnvbio_1_1io_1_1_output_file.html#a4af200a57c8a6aa36b4cfa96b49eb1a0">read_data_1</a>, <a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> <a class="el" href="group___sequence_i_o.html#gaeda29063a8a35631e18372d8a6001e27">io::SequenceDataHost</a> *<a class="el" href="structnvbio_1_1io_1_1_output_file.html#a47ca3d557b6408588ed4eab922e6f866">read_data_2</a>=NULL)</td></tr>
<tr class="separator:a637d0c9547a1c3e3aaa0c5aa2d2d3827"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a97bc4b3558dba005cff6b3a3cf73da5e"><td class="memItemLeft" align="right" valign="top">virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structnvbio_1_1io_1_1_output_file.html#a97bc4b3558dba005cff6b3a3cf73da5e">process</a> (struct <a class="el" href="structnvbio_1_1io_1_1_g_p_u_output_batch.html">GPUOutputBatch</a> &amp;gpu_batch, <a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> <a class="el" href="group___output.html#ga33a303d7df8d2b996a60298573bf109c">AlignmentMate</a> alignment_mate, <a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> <a class="el" href="group___output.html#ga14685a9d6ccdc4e08dd5454ded3821ae">AlignmentScore</a> alignment_score)</td></tr>
<tr class="separator:a97bc4b3558dba005cff6b3a3cf73da5e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0c7f1c3ab1ae5f1b975baab07806f395"><td class="memItemLeft" align="right" valign="top">virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structnvbio_1_1io_1_1_output_file.html#a0c7f1c3ab1ae5f1b975baab07806f395">end_batch</a> (void)</td></tr>
<tr class="memdesc:a0c7f1c3ab1ae5f1b975baab07806f395"><td class="mdescLeft">&#160;</td><td class="mdescRight">Mark a batch of alignment results as complete.  <a href="#a0c7f1c3ab1ae5f1b975baab07806f395">More...</a><br/></td></tr>
<tr class="separator:a0c7f1c3ab1ae5f1b975baab07806f395"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aab6138e6f0a387074f85f3a72b98a3dc"><td class="memItemLeft" align="right" valign="top">virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structnvbio_1_1io_1_1_output_file.html#aab6138e6f0a387074f85f3a72b98a3dc">close</a> (void)</td></tr>
<tr class="memdesc:aab6138e6f0a387074f85f3a72b98a3dc"><td class="mdescLeft">&#160;</td><td class="mdescRight">Flush and close the output file.  <a href="#aab6138e6f0a387074f85f3a72b98a3dc">More...</a><br/></td></tr>
<tr class="separator:aab6138e6f0a387074f85f3a72b98a3dc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8710fd9dd809dd8ba22c871c54f7b6c2"><td class="memItemLeft" align="right" valign="top">virtual <a class="el" href="structnvbio_1_1io_1_1_i_o_stats.html">IOStats</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structnvbio_1_1io_1_1_output_file.html#a8710fd9dd809dd8ba22c871c54f7b6c2">get_aggregate_statistics</a> (void)</td></tr>
<tr class="memdesc:a8710fd9dd809dd8ba22c871c54f7b6c2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns aggregate I/O statistics for this object.  <a href="#a8710fd9dd809dd8ba22c871c54f7b6c2">More...</a><br/></td></tr>
<tr class="separator:a8710fd9dd809dd8ba22c871c54f7b6c2"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-methods"></a>
Static Public Methods</h2></td></tr>
<tr class="memitem:af28e4dd0e708422eb0e29b0b11f02f2d"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="structnvbio_1_1io_1_1_output_file.html">OutputFile</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structnvbio_1_1io_1_1_output_file.html#af28e4dd0e708422eb0e29b0b11f02f2d">open</a> (<a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> char *<a class="el" href="structnvbio_1_1io_1_1_output_file.html#a81253ab272138730142146c13868a6ba">file_name</a>, <a class="el" href="group___output.html#ga1e89e8523b7e71ab8202a0821d7aa209">AlignmentType</a> aln_type, <a class="el" href="structnvbio_1_1io_1_1_b_n_t.html">BNT</a> <a class="el" href="structnvbio_1_1io_1_1_output_file.html#ae60d819329f6e0c7bb8b680a9f167e24">bnt</a>)</td></tr>
<tr class="separator:af28e4dd0e708422eb0e29b0b11f02f2d"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pro-methods"></a>
Proteced Methods</h2></td></tr>
<tr class="memitem:a176f305ce4fbf5900c2d5d85233ed1af"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structnvbio_1_1io_1_1_output_file.html#a176f305ce4fbf5900c2d5d85233ed1af">OutputFile</a> (<a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> char *<a class="el" href="structnvbio_1_1io_1_1_output_file.html#a81253ab272138730142146c13868a6ba">file_name</a>, <a class="el" href="group___output.html#ga1e89e8523b7e71ab8202a0821d7aa209">AlignmentType</a> <a class="el" href="structnvbio_1_1io_1_1_output_file.html#a5a3da8471750f654ffd87f692ea35552">alignment_type</a>, <a class="el" href="structnvbio_1_1io_1_1_b_n_t.html">BNT</a> <a class="el" href="structnvbio_1_1io_1_1_output_file.html#ae60d819329f6e0c7bb8b680a9f167e24">bnt</a>)</td></tr>
<tr class="separator:a176f305ce4fbf5900c2d5d85233ed1af"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0cc1fb9d07d75db430f16cf1eebed8ad"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structnvbio_1_1io_1_1_output_file.html#a0cc1fb9d07d75db430f16cf1eebed8ad">readback</a> (struct <a class="el" href="structnvbio_1_1io_1_1_c_p_u_output_batch.html">CPUOutputBatch</a> &amp;cpu_batch, <a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> struct <a class="el" href="structnvbio_1_1io_1_1_g_p_u_output_batch.html">GPUOutputBatch</a> &amp;gpu_batch, <a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> <a class="el" href="group___output.html#ga33a303d7df8d2b996a60298573bf109c">AlignmentMate</a> alignment_mate, <a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> <a class="el" href="group___output.html#ga14685a9d6ccdc4e08dd5454ded3821ae">AlignmentScore</a> alignment_score)</td></tr>
<tr class="separator:a0cc1fb9d07d75db430f16cf1eebed8ad"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pro-attribs"></a>
Protected Members</h2></td></tr>
<tr class="memitem:a81253ab272138730142146c13868a6ba"><td class="memItemLeft" align="right" valign="top"><a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structnvbio_1_1io_1_1_output_file.html#a81253ab272138730142146c13868a6ba">file_name</a></td></tr>
<tr class="memdesc:a81253ab272138730142146c13868a6ba"><td class="mdescLeft">&#160;</td><td class="mdescRight">Name of the file we're writing.  <a href="#a81253ab272138730142146c13868a6ba">More...</a><br/></td></tr>
<tr class="separator:a81253ab272138730142146c13868a6ba"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5a3da8471750f654ffd87f692ea35552"><td class="memItemLeft" align="right" valign="top"><a class="el" href="group___output.html#ga1e89e8523b7e71ab8202a0821d7aa209">AlignmentType</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structnvbio_1_1io_1_1_output_file.html#a5a3da8471750f654ffd87f692ea35552">alignment_type</a></td></tr>
<tr class="memdesc:a5a3da8471750f654ffd87f692ea35552"><td class="mdescLeft">&#160;</td><td class="mdescRight">The type of alignment we're running (single or paired-end)  <a href="#a5a3da8471750f654ffd87f692ea35552">More...</a><br/></td></tr>
<tr class="separator:a5a3da8471750f654ffd87f692ea35552"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae60d819329f6e0c7bb8b680a9f167e24"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structnvbio_1_1io_1_1_b_n_t.html">BNT</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structnvbio_1_1io_1_1_output_file.html#ae60d819329f6e0c7bb8b680a9f167e24">bnt</a></td></tr>
<tr class="memdesc:ae60d819329f6e0c7bb8b680a9f167e24"><td class="mdescLeft">&#160;</td><td class="mdescRight">Reference genome data handle.  <a href="#ae60d819329f6e0c7bb8b680a9f167e24">More...</a><br/></td></tr>
<tr class="separator:ae60d819329f6e0c7bb8b680a9f167e24"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab93d4b116920aea44923c634f7f79a13"><td class="memItemLeft" align="right" valign="top"><a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> <a class="el" href="structnvbio_1_1io_1_1_map_q_evaluator.html">io::MapQEvaluator</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structnvbio_1_1io_1_1_output_file.html#ab93d4b116920aea44923c634f7f79a13">mapq_evaluator</a></td></tr>
<tr class="memdesc:ab93d4b116920aea44923c634f7f79a13"><td class="mdescLeft">&#160;</td><td class="mdescRight">The MapQ evaluator struct used to compute mapping qualities during output.  <a href="#ab93d4b116920aea44923c634f7f79a13">More...</a><br/></td></tr>
<tr class="separator:ab93d4b116920aea44923c634f7f79a13"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aac600141218c722d6f69dfe624f85fb6"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structnvbio_1_1io_1_1_output_file.html#aac600141218c722d6f69dfe624f85fb6">mapq_filter</a></td></tr>
<tr class="memdesc:aac600141218c722d6f69dfe624f85fb6"><td class="mdescLeft">&#160;</td><td class="mdescRight">The current mapping quality filter: reads with a mapq below this value will be marked as not aligned.  <a href="#aac600141218c722d6f69dfe624f85fb6">More...</a><br/></td></tr>
<tr class="separator:aac600141218c722d6f69dfe624f85fb6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4af200a57c8a6aa36b4cfa96b49eb1a0"><td class="memItemLeft" align="right" valign="top"><a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> <a class="el" href="group___sequence_i_o.html#gaeda29063a8a35631e18372d8a6001e27">io::SequenceDataHost</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structnvbio_1_1io_1_1_output_file.html#a4af200a57c8a6aa36b4cfa96b49eb1a0">read_data_1</a></td></tr>
<tr class="separator:a4af200a57c8a6aa36b4cfa96b49eb1a0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a47ca3d557b6408588ed4eab922e6f866"><td class="memItemLeft" align="right" valign="top"><a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> <a class="el" href="group___sequence_i_o.html#gaeda29063a8a35631e18372d8a6001e27">io::SequenceDataHost</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structnvbio_1_1io_1_1_output_file.html#a47ca3d557b6408588ed4eab922e6f866">read_data_2</a></td></tr>
<tr class="separator:a47ca3d557b6408588ed4eab922e6f866"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0cf222c04199d6951ff00445237d2f4e"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structnvbio_1_1io_1_1_i_o_stats.html">IOStats</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structnvbio_1_1io_1_1_output_file.html#a0cf222c04199d6951ff00445237d2f4e">iostats</a></td></tr>
<tr class="memdesc:a0cf222c04199d6951ff00445237d2f4e"><td class="mdescLeft">&#160;</td><td class="mdescRight">I/O statistics.  <a href="#a0cf222c04199d6951ff00445237d2f4e">More...</a><br/></td></tr>
<tr class="separator:a0cf222c04199d6951ff00445237d2f4e"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a class="anchor" id="a176f305ce4fbf5900c2d5d85233ed1af"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">nvbio::io::OutputFile::OutputFile </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> char *&#160;</td>
          <td class="paramname"><em>file_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="group___output.html#ga1e89e8523b7e71ab8202a0821d7aa209">AlignmentType</a>&#160;</td>
          <td class="paramname"><em>alignment_type</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structnvbio_1_1io_1_1_b_n_t.html">BNT</a>&#160;</td>
          <td class="paramname"><em>bnt</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="output__file_8cpp_source.html#l00027">27</a> of file <a class="el" href="output__file_8cpp_source.html">output_file.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="adec2613fac91d4c2dc3a2a401b7b2e7d"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">nvbio::io::OutputFile::~OutputFile </td>
          <td>(</td>
          <td class="paramname">)</td><td></td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="output__file_8cpp_source.html#l00038">38</a> of file <a class="el" href="output__file_8cpp_source.html">output_file.cpp</a>.</p>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="aab6138e6f0a387074f85f3a72b98a3dc"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void nvbio::io::OutputFile::close </td>
          <td>(</td>
          <td class="paramtype">void&#160;</td>
          <td class="paramname">)</td><td></td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Flush and close the output file. </p>

<p>Reimplemented in <a class="el" href="structnvbio_1_1io_1_1_debug_output.html#a39501106b6a83852e5816844c648d86f">nvbio::io::DebugOutput</a>, <a class="el" href="structnvbio_1_1io_1_1_sam_output.html#abbfe3f26540e1de852563d190e84d2d1">nvbio::io::SamOutput</a>, and <a class="el" href="structnvbio_1_1io_1_1_bam_output.html#a98f5668632a34986126465caa0123d09">nvbio::io::BamOutput</a>.</p>

<p>Definition at line <a class="el" href="output__file_8cpp_source.html#l00070">70</a> of file <a class="el" href="output__file_8cpp_source.html">output_file.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="aa109737d25afa8d93a55572803aca499"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void nvbio::io::OutputFile::configure_mapq_evaluator </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> <a class="el" href="structnvbio_1_1io_1_1_map_q_evaluator.html">io::MapQEvaluator</a> *&#160;</td>
          <td class="paramname"><em>mapq</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>mapq_filter</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Configure the MapQ evaluator. Must be called prior to any batch processing. </p>

<p>Definition at line <a class="el" href="output__file_8cpp_source.html#l00042">42</a> of file <a class="el" href="output__file_8cpp_source.html">output_file.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a0c7f1c3ab1ae5f1b975baab07806f395"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void nvbio::io::OutputFile::end_batch </td>
          <td>(</td>
          <td class="paramtype">void&#160;</td>
          <td class="paramname">)</td><td></td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Mark a batch of alignment results as complete. </p>

<p>Reimplemented in <a class="el" href="structnvbio_1_1io_1_1_debug_output.html#a4d7b8684b358c16e3e2e45ba796586bd">nvbio::io::DebugOutput</a>, <a class="el" href="structnvbio_1_1io_1_1_sam_output.html#afdec1fa44f89c272c6f5ddc80cf657f7">nvbio::io::SamOutput</a>, and <a class="el" href="structnvbio_1_1io_1_1_bam_output.html#a3240184902a0bf5f336eef37cb360a28">nvbio::io::BamOutput</a>.</p>

<p>Definition at line <a class="el" href="output__file_8cpp_source.html#l00063">63</a> of file <a class="el" href="output__file_8cpp_source.html">output_file.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a8710fd9dd809dd8ba22c871c54f7b6c2"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structnvbio_1_1io_1_1_i_o_stats.html">IOStats</a> &amp; nvbio::io::OutputFile::get_aggregate_statistics </td>
          <td>(</td>
          <td class="paramtype">void&#160;</td>
          <td class="paramname">)</td><td></td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Returns aggregate I/O statistics for this object. </p>

<p>Definition at line <a class="el" href="output__file_8cpp_source.html#l00074">74</a> of file <a class="el" href="output__file_8cpp_source.html">output_file.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="af28e4dd0e708422eb0e29b0b11f02f2d"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structnvbio_1_1io_1_1_output_file.html">OutputFile</a> * nvbio::io::OutputFile::open </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> char *&#160;</td>
          <td class="paramname"><em>file_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="group___output.html#ga1e89e8523b7e71ab8202a0821d7aa209">AlignmentType</a>&#160;</td>
          <td class="paramname"><em>aln_type</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structnvbio_1_1io_1_1_b_n_t.html">BNT</a>&#160;</td>
          <td class="paramname"><em>bnt</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Factory method to create <a class="el" href="structnvbio_1_1io_1_1_output_file.html">OutputFile</a> objects </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[in]</td><td class="paramname">file_name</td><td>The name of the file to create (will be silently overwritten if it already exists). This method parses out the extension from the file name to determine what kind of file format to write. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">aln_type</td><td>The type of alignment (single or paired-end) </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">bnt</td><td>A handle to the reference genome </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A pointer to an <a class="el" href="structnvbio_1_1io_1_1_output_file.html">OutputFile</a> object, or NULL if an error occurs. </dd></dl>

<p>Definition at line <a class="el" href="output__file_8cpp_source.html#l00122">122</a> of file <a class="el" href="output__file_8cpp_source.html">output_file.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a97bc4b3558dba005cff6b3a3cf73da5e"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void nvbio::io::OutputFile::process </td>
          <td>(</td>
          <td class="paramtype">struct <a class="el" href="structnvbio_1_1io_1_1_g_p_u_output_batch.html">GPUOutputBatch</a> &amp;&#160;</td>
          <td class="paramname"><em>gpu_batch</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> <a class="el" href="group___output.html#ga33a303d7df8d2b996a60298573bf109c">AlignmentMate</a>&#160;</td>
          <td class="paramname"><em>alignment_mate</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> <a class="el" href="group___output.html#ga14685a9d6ccdc4e08dd5454ded3821ae">AlignmentScore</a>&#160;</td>
          <td class="paramname"><em>alignment_score</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Process a set of alignment results for the current batch. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">gpu_batch</td><td>Handle to the GPU buffers containing the alignment results </td></tr>
    <tr><td class="paramname">alignment_mate</td><td>Identifies the mate for this pass </td></tr>
    <tr><td class="paramname">alignment_score</td><td>Identifies the score type being calculated in this pass (currently either best or second-best) </td></tr>
  </table>
  </dd>
</dl>

<p>Reimplemented in <a class="el" href="structnvbio_1_1io_1_1_debug_output.html#a01d03e0536263d3fd66d06303ff3de43">nvbio::io::DebugOutput</a>, <a class="el" href="structnvbio_1_1io_1_1_sam_output.html#a64d34b17059c6361d6591cbf2b09286c">nvbio::io::SamOutput</a>, and <a class="el" href="structnvbio_1_1io_1_1_bam_output.html#aec05d8195fe5d7a77ae8bf22d630a2f5">nvbio::io::BamOutput</a>.</p>

<p>Definition at line <a class="el" href="output__file_8cpp_source.html#l00056">56</a> of file <a class="el" href="output__file_8cpp_source.html">output_file.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a0cc1fb9d07d75db430f16cf1eebed8ad"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void nvbio::io::OutputFile::readback </td>
          <td>(</td>
          <td class="paramtype">struct <a class="el" href="structnvbio_1_1io_1_1_c_p_u_output_batch.html">CPUOutputBatch</a> &amp;&#160;</td>
          <td class="paramname"><em>cpu_batch</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> struct <a class="el" href="structnvbio_1_1io_1_1_g_p_u_output_batch.html">GPUOutputBatch</a> &amp;&#160;</td>
          <td class="paramname"><em>gpu_batch</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> <a class="el" href="group___output.html#ga33a303d7df8d2b996a60298573bf109c">AlignmentMate</a>&#160;</td>
          <td class="paramname"><em>alignment_mate</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> <a class="el" href="group___output.html#ga14685a9d6ccdc4e08dd5454ded3821ae">AlignmentScore</a>&#160;</td>
          <td class="paramname"><em>alignment_score</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Read back batch data into the host </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[out]</td><td class="paramname">cpu_batch</td><td>The <a class="el" href="structnvbio_1_1io_1_1_c_p_u_output_batch.html">CPUOutputBatch</a> struct which will receive the data </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">gpu_batch</td><td>The GPU memory handle to read from </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">alignment_mate</td><td>Identifies the mate that the data in gpu_batch refers to </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">alignment_score</td><td>The type of score data in gpu_batch </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="output__file_8cpp_source.html#l00079">79</a> of file <a class="el" href="output__file_8cpp_source.html">output_file.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a637d0c9547a1c3e3aaa0c5aa2d2d3827"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void nvbio::io::OutputFile::start_batch </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> <a class="el" href="group___sequence_i_o.html#gaeda29063a8a35631e18372d8a6001e27">io::SequenceDataHost</a> *&#160;</td>
          <td class="paramname"><em>read_data_1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> <a class="el" href="group___sequence_i_o.html#gaeda29063a8a35631e18372d8a6001e27">io::SequenceDataHost</a> *&#160;</td>
          <td class="paramname"><em>read_data_2</em> = <code>NULL</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">virtual</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Begin a new batch of alignment results </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">read_data_1</td><td>The (host-side) read data pointer for the first mate </td></tr>
    <tr><td class="paramname">read_data_2</td><td>The (host-side) read data pointer for the second mate, if any (can be NULL for single-end alignment) </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="output__file_8cpp_source.html#l00048">48</a> of file <a class="el" href="output__file_8cpp_source.html">output_file.cpp</a>.</p>

</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a class="anchor" id="a5a3da8471750f654ffd87f692ea35552"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="group___output.html#ga1e89e8523b7e71ab8202a0821d7aa209">AlignmentType</a> nvbio::io::OutputFile::alignment_type</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>The type of alignment we're running (single or paired-end) </p>

<p>Definition at line <a class="el" href="output__file_8h_source.html#l00111">111</a> of file <a class="el" href="output__file_8h_source.html">output_file.h</a>.</p>

</div>
</div>
<a class="anchor" id="ae60d819329f6e0c7bb8b680a9f167e24"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structnvbio_1_1io_1_1_b_n_t.html">BNT</a> nvbio::io::OutputFile::bnt</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Reference genome data handle. </p>

<p>Definition at line <a class="el" href="output__file_8h_source.html#l00113">113</a> of file <a class="el" href="output__file_8h_source.html">output_file.h</a>.</p>

</div>
</div>
<a class="anchor" id="a81253ab272138730142146c13868a6ba"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> char* nvbio::io::OutputFile::file_name</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Name of the file we're writing. </p>

<p>Definition at line <a class="el" href="output__file_8h_source.html#l00109">109</a> of file <a class="el" href="output__file_8h_source.html">output_file.h</a>.</p>

</div>
</div>
<a class="anchor" id="a0cf222c04199d6951ff00445237d2f4e"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structnvbio_1_1io_1_1_i_o_stats.html">IOStats</a> nvbio::io::OutputFile::iostats</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>I/O statistics. </p>

<p>Definition at line <a class="el" href="output__file_8h_source.html#l00126">126</a> of file <a class="el" href="output__file_8h_source.html">output_file.h</a>.</p>

</div>
</div>
<a class="anchor" id="ab93d4b116920aea44923c634f7f79a13"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> <a class="el" href="structnvbio_1_1io_1_1_map_q_evaluator.html">io::MapQEvaluator</a>* nvbio::io::OutputFile::mapq_evaluator</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>The MapQ evaluator struct used to compute mapping qualities during output. </p>

<p>Definition at line <a class="el" href="output__file_8h_source.html#l00116">116</a> of file <a class="el" href="output__file_8h_source.html">output_file.h</a>.</p>

</div>
</div>
<a class="anchor" id="aac600141218c722d6f69dfe624f85fb6"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int nvbio::io::OutputFile::mapq_filter</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>The current mapping quality filter: reads with a mapq below this value will be marked as not aligned. </p>

<p>Definition at line <a class="el" href="output__file_8h_source.html#l00118">118</a> of file <a class="el" href="output__file_8h_source.html">output_file.h</a>.</p>

</div>
</div>
<a class="anchor" id="a4af200a57c8a6aa36b4cfa96b49eb1a0"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> <a class="el" href="group___sequence_i_o.html#gaeda29063a8a35631e18372d8a6001e27">io::SequenceDataHost</a>* nvbio::io::OutputFile::read_data_1</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Host-side copies of the read data for the current batch. These are set by start_batch and invalidated by end_batch. </p>

<p>Definition at line <a class="el" href="output__file_8h_source.html#l00122">122</a> of file <a class="el" href="output__file_8h_source.html">output_file.h</a>.</p>

</div>
</div>
<a class="anchor" id="a47ca3d557b6408588ed4eab922e6f866"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="vs_2zlib_2zconf_8h.html#a2c212835823e3c54a8ab6d95c652660e">const</a> <a class="el" href="group___sequence_i_o.html#gaeda29063a8a35631e18372d8a6001e27">io::SequenceDataHost</a>* nvbio::io::OutputFile::read_data_2</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Definition at line <a class="el" href="output__file_8h_source.html#l00123">123</a> of file <a class="el" href="output__file_8h_source.html">output_file.h</a>.</p>

</div>
</div>
<hr/>The documentation for this struct was generated from the following files:<ul>
<li>E:/p4sw/sw/gpgpu/nvbio/main/nvbio/io/output/<a class="el" href="output__file_8h_source.html">output_file.h</a></li>
<li>E:/p4sw/sw/gpgpu/nvbio/main/nvbio/io/output/<a class="el" href="output__file_8cpp_source.html">output_file.cpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on Wed Sep 10 2014 07:43:11 for NVBIO by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.4
</small></address>
</body>
</html>
